### DISCUSSION TITLE
Is there any way to make ripgrep return exit code 0 on no match?

### DISCUSSION THREAD

**Discussion by rew1nter** [BurntSushi/ripgrep#2675] [Q&A] [answered]
Hi there. 
I was executing ripgrep from code and found out this behavior. Is it possible to change the return code? if not what's the recommended solution?

**Comment by BurntSushi** [comment]
Short answer is that no, it isn't possible, and this is by design. In terms of what a recommend solution is, well, in order to say that, I need to know what the problem is. :-) Presuming you're writing a shell script, you might do `rg foo || true`. You might also use `rg` in an `if` statement. Honestly, there are a lot of different ways to approach it because it depends on what you're doing. And it's all pretty standard behavior because ripgrep behaves like grep here: exit with `0` when a match is found, `1` when no match is found and `2` when some other error occurs.

Probably this issue will give you the longer answer: https://github.com/BurntSushi/ripgrep/issues/2500

**Reply by rew1nter** [reply]
This is what I'm executing with `cat file | rg -v "(\[4\d{2}\]|\[3\d{2}\])" > data.txt`. Here file may be empty or have no match

> ripgrep behaves like grep here

Weird. As ripgrep was causing this error, I switched to grep for the time being cause it wasn't doing this :/

**Reply by BurntSushi** [reply]
Can you please provide a more specific reproduction? Can you also say what is actually going wrong? For example:

```
$ mkdir -p /tmp/rg2675
$ cd /tmp/rg2675
$ touch file
$ cat file | rg -v "(\[4\d{2}\]|\[3\d{2}\])" > data.txt
$ echo $?
1
$ cat file | grep -E -v "(\[4\d{2}\]|\[3\d{2}\])" > data.txt
grep: warning: stray \ before d
grep: warning: stray \ before d
$ cat file | grep -E -v "(\[4[0-9]{2}\]|\[3[0-9]{2}\])" > data.txt
$ echo $?
1
```

Everything above looks like it's behaving as expected. And ripgrep and grep behave the same (modulo the fact that grep doesn't support `\d`).

**Reply by rew1nter** [reply]
Alrighty. Here's the input file:
```
https://docs-dev.corp.google.com [302] [https://login.corp.google.com/request?s=docs-dev.corp.google.com:443/uberproxy/&d=https://docs-dev.corp.google.com/%3Fupxsrf%3DADM2IpRSKWoNJATW6bUrttPKMSp4lgLbGG_tf14D11XqYK1UrQ:1701788940589&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]
https://frame.ds-next-external-devel.qa.adz.google.com [302] [https://login.corp.google.com/request?s=frame.ds-next-external-devel.qa.adz.google.com:443/uberproxy/&d=https://frame.ds-next-external-devel.qa.adz.google.com/%3Fupxsrf%3DADM2IpRa4bJ9bSKUSiIjKNnRS4muL9QmtAYJvXL_hd3daKZt-Q:1701788940992&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]

```

Here's how I'm running it:
```go
package main

import (
	"log"
	"os"
	"os/exec"
)

func main() {
	cmd := exec.Command(
		"bash",
		"-c",
		`cat test | rg -v '(\[4\d{2}\]|\[3\d{2}\])'`,
	)
	cmd.Stderr = os.Stderr
	cmd.Stdout = os.Stdout
	if err := cmd.Run(); err != nil {
		log.Fatal(err)
	}
}

```
You can `go run file.go` to run this. See that it errors out with exit status 1. But replacing `rg` with `grep` causes no error

**Reply by BurntSushi** [reply]
What is the expected result? Please try to simplify things here. Your regex pattern has escapes in places, and it's actually ambiguous to me what the intended result should be. Do you expect that ripgrep shouldn't match? And when you replace `rg` with `grep`, is the exit code the only difference?

**Reply by BurntSushi** [reply]
For example:

```
$ grep -v '(\[4[0-9]{2}\]|\[3[0-9]{2}\])' test
https://docs-dev.corp.google.com [302] [https://login.corp.google.com/request?s=docs-dev.corp.google.com:443/uberproxy/&d=https://docs-dev.corp.google.com/%3Fupxsrf%3DADM2IpRSKWoNJATW6bUrttPKMSp4lgLbGG_tf14D11XqYK1UrQ:1701788940589&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]
https://frame.ds-next-external-devel.qa.adz.google.com [302] [https://login.corp.google.com/request?s=frame.ds-next-external-devel.qa.adz.google.com:443/uberproxy/&d=https://frame.ds-next-external-devel.qa.adz.google.com/%3Fupxsrf%3DADM2IpRa4bJ9bSKUSiIjKNnRS4muL9QmtAYJvXL_hd3daKZt-Q:1701788940992&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]

$ grep -E -v '(\[4[0-9]{2}\]|\[3[0-9]{2}\])' test

$ echo $?
1
```

I'm starting to think that the exit code is red herring here, and what you're _actually_ asking about is why ripgrep isn't matching while grep is. But I can't make heads or tails of what you're actually trying to do here.

**Reply by rew1nter** [reply]
> modulo the fact that grep doesn't support \d

Okay I just realized that this causes grep to return a match and ripgrep doesn't return any match. So both ripgrep and grep is behaving the same except for the fact that grep doesn't recognize `\d`

> What is the expected result?

I want ripgrep to silently pass with exit status 0 when no match is found.

**Reply by BurntSushi** [reply, marked as answer]
Maybe this is what you want?

```
$ grep -E -v '([4[0-9]{2}]\|[3[0-9]{2}])' test
https://docs-dev.corp.google.com [302] [https://login.corp.google.com/request?s=docs-dev.corp.google.com:443/uberproxy/&d=https://docs-dev.corp.google.com/%3Fupxsrf%3DADM2IpRSKWoNJATW6bUrttPKMSp4lgLbGG_tf14D11XqYK1UrQ:1701788940589&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]
https://frame.ds-next-external-devel.qa.adz.google.com [302] [https://login.corp.google.com/request?s=frame.ds-next-external-devel.qa.adz.google.com:443/uberproxy/&d=https://frame.ds-next-external-devel.qa.adz.google.com/%3Fupxsrf%3DADM2IpRa4bJ9bSKUSiIjKNnRS4muL9QmtAYJvXL_hd3daKZt-Q:1701788940992&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]

$ echo $?
0

$ rg -v '([4[0-9]{2}]\|[3[0-9]{2}])' test
1:https://docs-dev.corp.google.com [302] [https://login.corp.google.com/request?s=docs-dev.corp.google.com:443/uberproxy/&d=https://docs-dev.corp.google.com/%3Fupxsrf%3DADM2IpRSKWoNJATW6bUrttPKMSp4lgLbGG_tf14D11XqYK1UrQ:1701788940589&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]
2:https://frame.ds-next-external-devel.qa.adz.google.com [302] [https://login.corp.google.com/request?s=frame.ds-next-external-devel.qa.adz.google.com:443/uberproxy/&d=https://frame.ds-next-external-devel.qa.adz.google.com/%3Fupxsrf%3DADM2IpRa4bJ9bSKUSiIjKNnRS4muL9QmtAYJvXL_hd3daKZt-Q:1701788940992&maxAge=1200&authLevel=2000000&keyIds=M9V,bLf,DU5,2O9&c=1] [302 Moved] [HSTS]

$ echo $?
0
```

You opened this question asking about exit status codes, and ripgrep and grep should behave the same on that respect. But that does not mean that every single possible pattern has the same semantics in both ripgrep and grep. For example, `grep 'a|b'` and `grep -E 'a|b'` are two entirely different patterns. In the former case, `|` is matched literally, but in the latter case, `|` means "or" so it matches "a" or "b".

[ripgrep isn't a drop-in replacement for grep.](https://github.com/BurntSushi/ripgrep/blob/master/FAQ.md#posix4ever) Its exit status behavior should be the same, but the patterns each supports are not.

**Reply by BurntSushi** [reply]
> I want ripgrep to silently pass with exit status 0 when no match is found.

grep doesn't do that either, as I demonstrated above. If you can provide a _precise_ reproduction where this is not the case, I'd be happy to take a look. But I haven't seen one yet.

Otherwise, add `|| true` to the end of your shell command, as I said initially:

```go
package main

import (
	"log"
	"os"
	"os/exec"
)

func main() {
	cmd := exec.Command(
		"bash",
		"-c",
		`cat test | rg -v '(\[4[0-9]{2}\]|\[3[0-9]{2}\])' || true`,
	)
	cmd.Stderr = os.Stderr
	cmd.Stdout = os.Stdout
	if err := cmd.Run(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by rew1nter** [reply]
> grep doesn't do that either, as I demonstrated above

Understood :)

> Maybe this is what you want?

Yep. This regex seems to have solved the problem `(\[4[0-9]{2}\]|\[3[0-9]{2}\])`.

> Otherwise, add || true to the end of your shell command,

I haven't understood the usage of this `|| true `.
I also don't understand why `[0-9]` and `\d` causes different behaviour here

**Reply by rew1nter** [reply]
Btw thanks for helping me with this

**Reply by BurntSushi** [reply]
> I also don't understand why `[0-9]` and `\d` causes different behaviour here

My opinionated answer is that POSIX is to blame. Tools like grep are supposed to conform to a standard called POSIX, and POSIX exists to provide "portability." But in actual fact, POSIX is somewhat barebones on its own. So tools like, for example, GNU grep are said to be "POSIX compatible," but they actually provide _a lot_ more than what POSIX requires. Notice how the [POSIX specification for grep is tiny](https://pubs.opengroup.org/onlinepubs/009696899/utilities/grep.html).

The [POSIX specification for regexes](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html) actually does not define `\w`, `\s` or `\d` at all. Yet, your copy of grep almost certainly provides `\w` and `\s`. Yet, it doesn't provide `\d`. I have no clue why. You can use `grep -P` and get `\d`. For GNU grep anyway.

Interestingly, my copy of BSD grep on my mac machine does let you match `\d`:

```
$ grep -V
grep (BSD grep, GNU compatible) 2.6.0-FreeBSD

$ echo 5 | grep '\d'
5
```

Who woulda thunk it. Lol. This is what "POSIX compatibility" means. As long as you're very familiar with the specification, you can use tools in a way that is POSIX compatible by _artificially restricting yourself_ to what is specified. But as soon as you start doing stuff outside what is specified---and nobody will tell you when you do!---you enter into a land where portability is no longer guaranteed.

> I haven't understood the usage of this `|| true` .

It's part of your shell. It's a boolean expression. So if you do `cmd1 || cmd2 || ... || cmdN`, then `cmd1` will run, and if it succeeds, things stop. Otherwise, if it fails, `cmd2` is run. And so on. They are commonly found in `if` statements, but they can be used anywhere.

**Reply by BurntSushi** [reply]
Note that ripgrep supports `\d`, but it's Unicode aware. It will match every kind of digit, not just `[0-9]`. You can disable Unicode mode and write `(?-u:\d)`, which is equivalent to `[0-9]`.

**Reply by rew1nter** [reply]
Very interesting. This is a good read.
Learned something new about POSIX and grep today. 
So grep generally behaves differently on different POSIX operating systems. But I can use the `-P` flag with grep or ripgrep which supports `\d` and such
Understood the conditional as well. Thank you very much for taking your time to write this. 

Only a guy as cool as you could create ripgrep
